mln/tests/services/test_misc.py

```diff
@@ -85,6 +100,24 @@ def test(self):
 		self.assertFalse(self.user.inventory.filter(item_id=self.REQUIREMENT_ID).exists())
 		self.assertTrue(self.user.inventory.filter(item_id=self.ITEM_ID, qty=1).exists())
 
+class CraftMasterpiece_Ok(TestCase):
+	SETUP = has_masterpiece_blueprint, has_masterpiece_requirement
+	
+	def test(self):
+		use_blueprint(self.user, self.MASTERPIECE_BLUEPRINT_ID)
+		self.assertFalse(self.user.inventory.filter(item_id=self.MASTERPIECE_REQUIREMENT_ID).exists())
+		self.assertTrue(self.user.inventory.filter(item_id=self.MASTERPIECE_ITEM_ID, qty=1).exists())
+		self.assertTrue(self.user.profile.rank == 1)
+
+class CraftMasterpiece_Dupe(TestCase):
+	SETUP = has_masterpiece_blueprint, has_masterpiece_requirement, has_masterpiece
+	
+	def test(self):
+		use_blueprint(self.user, self.MASTERPIECE_BLUEPRINT_ID)
+		self.assertFalse(self.user.inventory.filter(item_id=self.MASTERPIECE_REQUIREMENT_ID).exists())
+		self.assertFalse(self.user.inventory.filter(item_id=self.MASTERPIECE_ITEM_ID, qty=2).exists())
+		self.assertFalse(self.user.profile.rank == 2)

```

Comment:
The presence of the item in the inventory alone would not increment the rank (only the crafting action would do that), so the rank would be expected to be zero in this case.
```suggestion
		self.assertEqual(self.user.profile.rank, 0)
```